### ISSUE TITLE
Best way to set CORSIKA seeds in order to assure reproducibility

### ISSUE THREAD

**Issue by RaulRPrado** [gammasim/gammasim-tools#121]
In corsika_config, the CORSIKA seeds should be set deterministically from the run parameters. By now, the seed of the python random module is determinaed deterministically and the CORSIKA seeds are determined from random.uniform. 
Is that enough to assure reproducibility?

**Comment by orelgueta** [comment]
Are the CORSIKA seeds later written out in the output where we can keep them in the provenance info?

**Comment by GernotMaier** [comment]
Yes, see the event header block attached (from the CORSIKA manual)

![Screenshot 2021-04-27 at 09 20 58](https://user-images.githubusercontent.com/14001688/116201543-e3b20000-a739-11eb-95d2-06381a636824.png)

**Comment by RaulRPrado** [comment]
> Are the CORSIKA seeds later written out in the output where we can keep them in the provenance info?

Yes, for each run, CORSIKA stores the actual input file that was used, including the seeds.
No need to read it from the output file.

**Comment by orelgueta** [comment]
OK, so in that case, as long as we keep the seeds that are used, I don't see an issue with the system you set up. We will need to allow for the option to provide specific seeds to CORSIKA instead of that, but otherwise it should be fine I think. Either way, 
100% reproducibility is very hard to achieve in the long run (running on different systems sometimes produces different results), but this seems close enough.

**Comment by GernotMaier** [comment]
Suggest to close this issue.

As discussed are the CORSIKA seeds saved in the output file. We also know about the limitations of re-running CORSIKA with exactly the same seeds due to limits of machine accuracy. Don't see anything else to be addressed in the context of this issue.

**Comment by orelgueta** [comment]
The current implementation actually uses the `corsika_autoinputs` mechanism to generate unique random seeds for CORSIKA. The ones produced in simtools are ignored in most cases. In either case we are safe from the point of view of using unique random seeds for each run (except for testing).

Regarding reproducibility, the log file prints out the random seeds, which could be enough I guess. However, the metaparameters contain the random seeds for sim_telarray. So, adding the CORSIKA random seeds would be the ideal solution I think. The metaparameters are available in the simulation file itself, so no need to look at any other file in order to obtain the full configuration to reproduce a run. 

If you think it's worth the effort, we can look into it and ask Konrad about implementation. If you think the chance of it being useful are too small, please feel free to close this issue.

**Comment by GernotMaier** [comment]
Agree, this should be part of the metaparameters. Let’s discuss this with Konrad.

**Comment by orelgueta** [comment]
Answer from Konrad about adding the CORSIKA seeds to the metaparameters:

> The difference between sim_telarray random seeds and those from
> CORSIKA is that the former are available in the configuration stage
> and the latter only after the configuration is complete and processing
> of data has started. At that time the metaparameters have already
> been emitted.
> 
> Where the CORSIKA random number seeds are available is in the
> copy of the CORSIKA inputs, data block type 1212, right after the
> run header and the MC run header blocks.

pyeventio reads the CORSIKA input cards data block, so we can get an array with the random seeds easily from any sim_telarray produced file by doing:

```
with SimTelFile(file) as f:
    corsika_seeds_re = re.findall(rb"SEED (\d+)", f.corsika_inputcards[0])
    corsika_seeds = list(map(int, corsika_seeds_re))
    print(corsika_seeds)

> [534, 220, 1104, 382]
```

I don't think we need to implement it anywhere, it is not foreseen to be a part of a simtools application. It is useful to have this piece of code here though in case it is necessary.

Regarding this specific issue, I suggest we keep it open a little longer and I will assign it to myself. The issue actually discusses the possibility to set those seeds manually. Right now for testing we can only set a predefined set of random seeds, but I can easily change it to accept a list of four random seeds. I will do that and then we close the issue.

**Comment by GernotMaier** [comment]
@orelgueta - could you check this again after the holidays and explain why we can't close this? I can't read any actionable item from above discussion.
